File: src/identity.py
"""Space identity model: one colour per Space, rendered as a coloured dot.

A Space's identity is a single **colour**. Everything visible is derived from it:

* the sidebar marker is a dot drawn in that colour;
* the chrome tint accent is that colour.

Herdr's sidebar token styling is *static config* -- there is no way to colour a
row from a metadata value -- so each palette colour gets its own pre-styled
metadata token (`$sd_blue`, `$sd_purple`, ...). All slot tokens sit in the row
with their fixed `fg`, and only the one matching a Space carries a value, so
exactly one coloured dot appears per row. This is the "statically configured
metadata tokens represent palette slots" approach.

Identity is keyed on the Herdr workspace ID, so renaming a Space never changes it.
"""

import re

# Eight clearly separable hues. Bright pure red and green are excluded so a Space
# colour is never confusable with Herdr's semantic agent status colours.
# Twelve pastel hues in the register dev tooling actually uses -- drawn from
# Catppuccin, Tokyo Night, One Dark, Rose Pine, Everforest and Dracula. Ordered by
# hue so the picker reads as a colour wheel.
#
# Pastels are *safer* than saturated colours here, not just prettier: every one of
# these sits >=110 perceptual units away from gruvbox's status red/green/yellow,
# where the previous saturated palette came as close as 80. A Space colour is
# therefore harder to mistake for an Agent state.
#
# The flip side is that 12 pastels are inherently less separable than 8 saturated
# ones (they share a narrow luminance band), so `allocate` picks by maximum
# perceptual distance from the colours already in use rather than round-robin.
PALETTE = [
    ("rose",     "#eba0ac"),
    ("peach",    "#fab387"),
    ("amber",    "#e5c07b"),
    ("sage",     "#a6d189"),
    ("aqua",     "#7fd6c1"),
    ("cyan",     "#8be9fd"),
    ("steel",    "#8ca0b3"),
    ("azure",    "#7aa2f7"),
    ("lavender", "#b4befe"),
    ("mauve",    "#cba6f7"),
    ("orchid",   "#e0a3e8"),
    ("blush",    "#f5c2e7"),
]

PALETTE_BY_NAME = dict(PALETTE)
SLOT_NAMES = [name for name, _ in PALETTE]

# Metadata token per palette slot. Must satisfy ^[A-Za-z0-9_-]{1,32}$.
TOKEN_PREFIX = "sd_"

# Default marker glyph. Single-width and unambiguous; override in settings.json
# (`marker`) with e.g. "◆", "■" or "▊" if it reads too close to your state icon.
DEFAULT_MARKER = "●"          # ●

HEX_RE = re.compile(r'^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$')


def slot_token(slot_name):
    return TOKEN_PREFIX + slot_name


def all_slot_tokens():
    return [slot_token(n) for n in SLOT_NAMES]


def normalise_hex(value):
    """Return a canonical #rrggbb string, or None if not a valid hex colour."""
    if not isinstance(value, str):
        return None
    v = value.strip()
    if not HEX_RE.match(v):
        return None
    v = v.lower()
    if len(v) == 4:
        v = "#" + "".join(c * 2 for c in v[1:])
    return v


def resolve_colour(value):
    """Accept a palette name or a hex string. -> #rrggbb or None."""
    if not isinstance(value, str):
        return None
    name = value.strip().lower()
    if name in PALETTE_BY_NAME:
        return PALETTE_BY_NAME[name]
    return normalise_hex(value)


def _rgb(hexv):
    v = hexv.lstrip("#")
    return int(v[0:2], 16), int(v[2:4], 16), int(v[4:6], 16)


def slot_for_colour(colour):
    """Palette slot whose styled token should carry the dot for `colour`.

    An exact palette colour maps to its own slot. A custom hex has no pre-styled
    token of its own, so it borrows the nearest palette slot for the dot while the
    exact hex is still used for the chrome tint. Documented, not silent.
    """
    target = normalise_hex(colour)
    if not target:
        return SLOT_NAMES[0]
    for name, hexv in PALETTE:
        if hexv == target:
            return name
    tr, tg, tb = _rgb(target)
    best, best_d = SLOT_NAMES[0], None
    for name, hexv in PALETTE:
        r, g, b = _rgb(hexv)
        # weighted to approximate perceptual distance
        d = 2 * (r - tr) ** 2 + 4 * (g - tg) ** 2 + 3 * (b - tb) ** 2
        if best_d is None or d < best_d:
            best, best_d = name, d
    return best


def is_exact_palette_colour(colour):
    return normalise_hex(colour) in {h for _n, h in PALETTE}


def colour_name(colour):
    """Palette name for an exact match, else the hex itself."""
    target = normalise_hex(colour)
    for name, hexv in PALETTE:
        if hexv == target:
            return name
    return target or "?"


def distance(a, b):
    """Weighted RGB distance -- a cheap perceptual approximation."""
    ah, bh = normalise_hex(a), normalise_hex(b)
    if not ah or not bh:
        return 0.0
    ar, ag, ab = _rgb(ah)
    br, bg, bb = _rgb(bh)
    rm = (ar + br) / 2.0
    return (((2 + rm / 256.0) * (ar - br) ** 2)
            + (4 * (ag - bg) ** 2)
            + ((2 + (255 - rm) / 256.0) * (ab - bb) ** 2)) ** 0.5


# Below this, two Space colours are close enough to be worth warning about.
CLOSE_ENOUGH = 80.0


def _ordered(workspaces):
    return sorted(workspaces, key=lambda w: (w.get("number") or 0,
                                             w.get("workspace_id") or ""))


def _neighbour_colours(workspaces, identities, workspace_id):
    order = [w.get("workspace_id") for w in _ordered(workspaces)]
    if workspace_id not in order:
        return set()
    i = order.index(workspace_id)
    out = set()
    for j in (i - 1, i + 1):
        if 0 <= j < len(order):
            ident = identities.get(order[j])
            if ident and ident.get("colour"):
                out.add(ident["colour"].lower())
    return out


def allocate(state, workspace_id, workspaces):
    """Pick the palette colour most visually distinct from the ones in use.

    Neighbouring Spaces weigh double, since those are the rows the eye compares.
    With 12 pastels and a handful of Spaces this reliably spreads across the wheel;
    only once colours run low does it start reusing them.
    """
    identities = state.get("identities") or {}
    live = {w.get("workspace_id") for w in workspaces}

    adjacent = _neighbour_colours(workspaces, identities, workspace_id)
    in_use = [(i.get("colour") or "").lower()
              for wid, i in identities.items()
              if wid != workspace_id and wid in live and i.get("colour")]

    best = None
    for name, hexv in PALETTE:
        if not in_use and not adjacent:
            score = 1e9 - PALETTE.index((name, hexv))   # stable first pick
        else:
            d_all = min([distance(hexv, c) for c in in_use] or [1e6])
            d_adj = min([distance(hexv, c) for c in adjacent] or [1e6])
            score = min(d_all, d_adj * 0.5) if adjacent else d_all
        if best is None or score > best[0]:
            best = (score, hexv)

    state["alloc_cursor"] = (int(state.get("alloc_cursor") or 0) + 1) % len(PALETTE)
    return {"colour": best[1], "origin": "auto"}


def close_pairs(state, workspaces):
    """Live Space pairs whose colours may be hard to tell apart."""
    identities = state.get("identities") or {}
    live = [w.get("workspace_id") for w in _ordered(workspaces)]
    out = []
    for i, a in enumerate(live):
        for b in live[i + 1:]:
            ca = (identities.get(a) or {}).get("colour")
            cb = (identities.get(b) or {}).get("colour")
            if not ca or not cb:
                continue
            d = distance(ca, cb)
            if d < CLOSE_ENOUGH:
                out.append((a, b, d))
    return sorted(out, key=lambda x: x[2])


def ensure_all(state, workspaces):
    """Give every known workspace an identity. Returns the list of new IDs."""
    added = []
    for w in _ordered(workspaces):
        wid = w.get("workspace_id")
        if not wid:
            continue
        existing = (state.get("identities") or {}).get(wid)
        if existing and existing.get("colour"):
            continue
        if existing:
            # migrate a legacy emoji-era entry that somehow lost its colour
            existing.update(allocate(state, wid, workspaces))
            continue
        state.setdefault("identities", {})[wid] = allocate(state, wid, workspaces)
        added.append(wid)
    return added

File: src/spaces/identity.ts
import { Predicate, Result, Schema } from "effect"

import type { PluginState } from "../state/store.ts"
import { PALETTE, SLOT_NAMES, paletteHex } from "./palette.ts"

type Json = typeof Schema.Json.Type

type JsonObject = typeof Schema.JsonObject.Type

export { DEFAULT_MARKER, PALETTE, SLOT_NAMES, allSlotTokens, slotToken } from "./palette.ts"

const HEX_RE = /^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/

export const CLOSE_ENOUGH = 80.0

export type WorkspaceSnapshot = {
  readonly workspace_id?: string
  readonly number?: number
  readonly label?: string
  readonly focused?: boolean
}

export type IdentityAssignment = {
  colour: string
  origin: string
}

type Rgb = {
  readonly r: number
  readonly g: number
  readonly b: number
}

export function normaliseHex(value: string): string | undefined {
  const trimmed = value.trim()

  if (!HEX_RE.test(trimmed)) return undefined

  const lower = trimmed.toLowerCase()

  if (lower.length === 4) {
    return `#${lower[1]}${lower[1]}${lower[2]}${lower[2]}${lower[3]}${lower[3]}`
  }

  return lower
}

export function resolveColour(value: string): string | undefined {
  const named = paletteHex(value.trim().toLowerCase())

  if (named !== undefined) return named

  return normaliseHex(value)
}

function rgb(hexv: string): Rgb {
  const v = hexv.replace("#", "")

  return {
    r: Number.parseInt(v.slice(0, 2), 16),
    g: Number.parseInt(v.slice(2, 4), 16),
    b: Number.parseInt(v.slice(4, 6), 16),
  }
}

export function slotForColour(colour: string): string {
  const target = normaliseHex(colour)
  const fallback = SLOT_NAMES[0] ?? "rose"

  if (target === undefined) return fallback

  for (const [name, hexv] of PALETTE) {
    if (hexv === target) return name
  }

  const wanted = rgb(target)
  let best = fallback
  let bestD: number | undefined

  for (const [name, hexv] of PALETTE) {
    const sample = rgb(hexv)

    const d = 2 * (sample.r - wanted.r) ** 2
      + 4 * (sample.g - wanted.g) ** 2
      + 3 * (sample.b - wanted.b) ** 2

    if (bestD === undefined || d < bestD) {
      best = name
      bestD = d
    }
  }

  return best
}

export function isExactPaletteColour(colour: string): boolean {
  const target = normaliseHex(colour)

  if (target === undefined) return false

  for (const [, hexv] of PALETTE) {
    if (hexv === target) return true
  }

  return false
}

export function colourName(colour: string): string {
  const target = normaliseHex(colour)

  for (const [name, hexv] of PALETTE) {
    if (hexv === target) return name
  }

  return target ?? "?"
}

export function distance(a: string, b: string): number {
  const ah = normaliseHex(a)
  const bh = normaliseHex(b)

  if (ah === undefined || bh === undefined) return 0.0

  const left = rgb(ah)
  const right = rgb(bh)
  const rm = (left.r + right.r) / 2.0

  return Math.sqrt(
    (2 + rm / 256.0) * (left.r - right.r) ** 2
      + 4 * (left.g - right.g) ** 2
      + (2 + (255 - rm) / 256.0) * (left.b - right.b) ** 2,
  )
}

function workspaceIdOf(workspace: WorkspaceSnapshot): string {
  return workspace.workspace_id ?? ""
}

function ordered(workspaces: readonly WorkspaceSnapshot[]): WorkspaceSnapshot[] {
  return [...workspaces].sort((left, right) => {
    const ln = left.number || 0
    const rn = right.number || 0

    if (ln !== rn) return ln - rn

    const lid = workspaceIdOf(left)
    const rid = workspaceIdOf(right)

    if (lid < rid) return -1

    if (lid > rid) return 1

    return 0
  })
}

function asJsonObject(value: Json): JsonObject | undefined {
  const decoded = Schema.decodeUnknownResult(Schema.JsonObject)(value)

  if (Result.isFailure(decoded)) return undefined

  return decoded.success
}

function identityColour(identities: PluginState["identities"], workspaceId: string): string | undefined {
  const raw = identities[workspaceId]

  if (raw === undefined) return undefined

  const info = asJsonObject(raw)

  if (info === undefined) return undefined

  const colour = info.colour

  if (!Predicate.isString(colour) || colour === "") return undefined

  return colour
}

function neighbourColours(
  workspaces: readonly WorkspaceSnapshot[],
  identities: PluginState["identities"],
  workspaceId: string,
): Set<string> {
  const order = ordered(workspaces).map(workspaceIdOf)
  const index = order.indexOf(workspaceId)

  if (index < 0) return new Set()

  const out = new Set<string>()

  for (const j of [index - 1, index + 1]) {
    const neighbourId = order[j]

    if (neighbourId === undefined) continue

    const colour = identityColour(identities, neighbourId)

    if (colour !== undefined) out.add(colour.toLowerCase())
  }

  return out
}

export function allocate(
  state: PluginState,
  workspaceId: string,
  workspaces: readonly WorkspaceSnapshot[],
): IdentityAssignment {
  const identities = state.identities
  const live = new Set<string>()

  for (const workspace of workspaces) {
    const id = workspaceIdOf(workspace)

    if (id !== "") live.add(id)
  }

  const adjacent = neighbourColours(workspaces, identities, workspaceId)
  const inUse: string[] = []

  for (const wid of Object.keys(identities)) {
    if (wid === workspaceId || !live.has(wid)) continue

    const colour = identityColour(identities, wid)

    if (colour !== undefined) inUse.push(colour.toLowerCase())
  }

  let bestHex: string = PALETTE[0]?.[1] ?? "#eba0ac"
  let bestScore: number | undefined

  for (let index = 0; index < PALETTE.length; index++) {
    const entry = PALETTE[index]

    if (entry === undefined) continue

    const hexv = entry[1]
    let score: number

    if (inUse.length === 0 && adjacent.size === 0) {
      score = 1e9 - index
    } else {
      const dAll = minDistance(hexv, inUse, 1e6)
      const dAdj = minDistance(hexv, adjacent, 1e6)

      score = adjacent.size > 0 ? Math.min(dAll, dAdj * 0.5) : dAll
    }

    if (bestScore === undefined || score > bestScore) {
      bestScore = score
      bestHex = hexv
    }
  }

  const cursor = Number(state.alloc_cursor) || 0

  state.alloc_cursor = (Math.trunc(cursor) + 1) % PALETTE.length

  return { colour: bestHex, origin: "auto" }
}

function minDistance(
  hexv: string,
  colours: Iterable<string>,
  empty: number,
): number {
  let best: number | undefined

  for (const colour of colours) {
    const d = distance(hexv, colour)

    if (best === undefined || d < best) best = d
  }

  return best ?? empty
}

export type ClosePair = readonly [string, string, number]

export function closePairs(
  state: PluginState,
  workspaces: readonly WorkspaceSnapshot[],
): ClosePair[] {
  const identities = state.identities
  const live = ordered(workspaces).map(workspaceIdOf)
  const out: ClosePair[] = []

  for (let i = 0; i < live.length; i++) {
    const a = live[i]

    if (a === undefined || a === "") continue

    for (let j = i + 1; j < live.length; j++) {
      const b = live[j]

      if (b === undefined || b === "") continue

      const ca = identityColour(identities, a)
      const cb = identityColour(identities, b)

      if (ca === undefined || cb === undefined) continue

      const d = distance(ca, cb)

      if (d < CLOSE_ENOUGH) out.push([a, b, d])
    }
  }

  return out.sort((left, right) => left[2] - right[2])
}

export function ensureAll(
  state: PluginState,
  workspaces: readonly WorkspaceSnapshot[],
): string[] {
  const added: string[] = []

  for (const workspace of ordered(workspaces)) {
    const wid = workspace.workspace_id

    if (wid === undefined || wid === "") continue

    const raw = state.identities[wid]
    const existing = raw === undefined ? undefined : asJsonObject(raw)

    if (existing !== undefined) {
      const colour = existing.colour

      if (Predicate.isString(colour) && colour !== "") continue

      const assigned = allocate(state, wid, workspaces)

      state.identities[wid] = { ...existing, colour: assigned.colour, origin: assigned.origin }
      continue
    }

    state.identities[wid] = allocate(state, wid, workspaces)
    added.push(wid)
  }

  return added
}
